### PR TITLE
Camelot V3

### PR DESCRIPTION
**Pull request by TechnoGeek01**
- [x] Testing

Some important points to be considered:-
* The tickspacing of pools is not immutable.
* If tickspacing is changed in a way that either of our tickUpper and tickLower are not divisible by this new tickspacing, users will not be able to mint new positions adhering to our farm ticks and they will also not be able to increase liquidity of their current positions.
* We can reduce risk by choosing tickLower and tickUpper as (highly divisible numbers and there should be more common divisors/factors between the two numbers) in our farm tick requirement. And ofcourse both the numbers should be divisible by the tickspacing of the pool at the time of farm launch.
* Camelot V3 pools can set tickspacing upto 500. 

Other differences from Uniswap V3 -> https://docs.algebra.finance/algebra-integral-documentation/algebra-integral-technical-reference/integration-process/migration-from-uniswapv3

### CODE REVIEW

File: contracts/e721-farms/camelotV3/CamelotV3Farm.sol
Comment (parv3213) [inline contracts/e721-farms/camelotV3/CamelotV3Farm.sol:1]: Can you remove fee-related checks from UniV3Farm as well?

File: test/e721-farms/camelotV3/CamelotV3Farm.t.sol [hunk 1]
```diff
@@ -0,0 +1,1041 @@
...
+        );
+        lockup
+            ? assertEq(
+                CamelotV3Farm(farm).getRewardFundInfo(CamelotV3Farm(farm).LOCKUP_FUND_ID()).totalLiquidity,
+                oldTotalFundLiquidity[0] + loggedActualLiquidity
+            )
+            : assert(true);
+    }
+
+    // Users will not be able to increaseLiquidity of their current positions if the pool’s tickSpacing is changed.
```
Comment (parv3213) [inline test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: What do we want to do with this test?

**Review by parv3213 — APPROVED** [review]

File: test/e721-farms/camelotV3/CamelotV3Farm.t.sol (hunk 1 above)
Comment (TechnoGeek01) [reply test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: I will remove it!

File: contracts/e721-farms/camelotV3/CamelotV3Farm.sol
Comment (parv3213) [reply contracts/e721-farms/camelotV3/CamelotV3Farm.sol:1]: Oh, I see you have done that in another PR. Alright 👍

Comment (TechnoGeek01) [reply contracts/e721-farms/camelotV3/CamelotV3Farm.sol:1]: I have worked on similar changes in this pr https://github.com/Sperax/Demeter-Protocol/pull/53

File: test/e721-farms/camelotV3/CamelotV3Farm.t.sol (hunk 1 above)
Comment (TechnoGeek01) [reply test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: fixed in [6db59ac](https://github.com/Sperax/Demeter-Protocol/pull/52/commits/6db59acf6d8f4b49118e4f03e4d0271c015c6c5e)

File: contracts/e721-farms/camelotV3/interfaces/ICamelotV3NonfungiblePositionManagerUtils.sol [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    int24 tickLower;
+    int24 tickUpper;
+    uint128 liquidity;
+    uint256 feeGrowthInside0LastX128;
+    uint256 feeGrowthInside1LastX128;
+    uint128 tokensOwed0;
+    uint128 tokensOwed1;
+}
+
+interface INFPMUtils {
```
Comment (parv3213) [inline contracts/e721-farms/camelotV3/interfaces/ICamelotV3NonfungiblePositionManagerUtils.sol]: Rename according to filename?

File: package.json [hunk 3]
```diff
@@ -22,6 +22,8 @@
     "forge-coverage": "forge coverage --report lcov && lcov --remove ./lcov.info -o ./lcov.info 'test/*' && rm -rf ./coverage && genhtml lcov.info --output-dir coverage && mv lcov.info ./coverage"
   },
   "dependencies": {
+    "@cryptoalgebra/v1.9-directional-fee-periphery": "^1.0.1",
+    "@cryptoalgebra/v1.9-periphery": "^1.0.1",
```
Comment (parv3213) [inline package.json]: Are you using both of these packages?

File: test/e721-farms/camelotV3/CamelotV3Farm.t.sol [hunk 4]
```diff
@@ -0,0 +1,980 @@
...
+            oldCommonTotalLiquidity - liquidityToWithdraw
+        );
+    }
+
+    function test_getTokenAmounts() public depositSetup(lockupFarm, true) {
+        // Manual testing
+        address[] memory tokens;
+        uint256[] memory amount;
+        (tokens, amount) = CamelotV3Farm(lockupFarm).getTokenAmounts();
+        emit log_named_uint("amount0", amount[0]);
```
Comment (parv3213) [inline test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: Instead of logging can we assert these?

File: package.json (hunk 3 above)
Comment (TechnoGeek01) [reply package.json]: only the directional fee package is required. Will remove it.

Comment (TechnoGeek01) [reply package.json]: fixed in [3060b15](https://github.com/Sperax/Demeter-Protocol/pull/52/commits/3060b1514c8f8f56196f4b09aca7dc9145e2bacb)

File: contracts/e721-farms/camelotV3/interfaces/ICamelotV3NonfungiblePositionManagerUtils.sol (hunk 2 above)
Comment (TechnoGeek01) [reply contracts/e721-farms/camelotV3/interfaces/ICamelotV3NonfungiblePositionManagerUtils.sol]: fixed in [1560fd6](https://github.com/Sperax/Demeter-Protocol/pull/52/commits/1560fd6600dcbd1aa5e684e113d4cc5a125fc134)

File: test/e721-farms/camelotV3/CamelotV3Farm.t.sol (hunk 4 above)
Comment (TechnoGeek01) [reply test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: These are complicated to assert as the price keeps changing as per market conditions.

Comment (parv3213) [reply test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: We must assert, logging is of no use. Maybe assert based on the formula used in getTokenAmounts?

Comment (TechnoGeek01) [reply test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: fixed in [a1d1879](https://github.com/Sperax/Demeter-Protocol/pull/52/commits/a1d187951c298576458f9dbfbeaf5c2ca1da3434)

File: test/e721-farms/camelotV3/CamelotV3Farm.t.sol [hunk 5]
```diff
@@ -0,0 +1,1030 @@
...
+        assertEq(CamelotV3Farm(farmProxy).camelotV3Factory(), CAMELOT_V3_FACTORY);
+        assertEq(CamelotV3Farm(farmProxy).nftContract(), NFPM);
+        assertEq(CamelotV3Farm(farmProxy).camelotUtils(), CAMELOT_V3_UTILS);
+        assertEq(CamelotV3Farm(farmProxy).nfpmUtils(), CAMELOT_V3_NONFUNGIBLE_POSITION_MANAGER_UTILS);
+    }
+}
+
+abstract contract OnERC721ReceivedTest is CamelotV3FarmTest {
+    function test_OnERC721Received_RevertWhen_IncorrectPoolToken() public useKnownActor(user) {
+        uint256 depositAmount1 = 1e3 * 10 ** ERC20(DAI).decimals();
```
Comment (YashP16) [inline test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: This chunk of code could be a single function that mints token and returns tokenId

File: test/e721-farms/camelotV3/CamelotV3Farm.t.sol [hunk 6]
```diff
@@ -0,0 +1,1030 @@
...
+            _camelotV3Factory: CAMELOT_V3_FACTORY,
+            _nftContract: NFPM,
+            _camelotUtils: CAMELOT_V3_UTILS,
+            _nfpmUtils: CAMELOT_V3_NONFUNGIBLE_POSITION_MANAGER_UTILS
+        });
+    }
+
+    function test_Initialize_RevertWhen_InvalidCamelotPoolConfig() public {
+        vm.expectRevert(abi.encodeWithSelector(CamelotV3Farm.InvalidCamelotPoolConfig.selector));
+        CamelotV3Farm(farmProxy).initialize({
```
Comment (YashP16) [inline test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: Update the comment to highlight the exact line that should fail

**Review by YashP16 — CHANGES_REQUESTED** [review]

File: test/e721-farms/camelotV3/CamelotV3Farm.t.sol (hunk 6 above)
Comment (TechnoGeek01) [reply test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: There were some redundant cases in this test case. Removed them.

fixed in [a36bd25](https://github.com/Sperax/Demeter-Protocol/pull/52/commits/a36bd2578e2d5bbb0f1a54a6fa04d99e9fa245d0)

Comment (TechnoGeek01) [reply test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: Same mistake done in Uniswap V3 test cases as well. Shall I remove them from there as well?

File: test/e721-farms/camelotV3/CamelotV3Farm.t.sol (hunk 5 above)
Comment (TechnoGeek01) [reply test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: fixed in [6929aef](https://github.com/Sperax/Demeter-Protocol/pull/52/commits/6929aef2c3bb0efcb59196a67bc0e201b30338e8)

Comment (TechnoGeek01) [reply test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: We can do this in UniV3 test cases as well. Shall I do this?

Comment (YashP16) [reply test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: Lets do it in another PR later

File: test/e721-farms/camelotV3/CamelotV3Farm.t.sol (hunk 6 above)
Comment (YashP16) [reply test/e721-farms/camelotV3/CamelotV3Farm.t.sol]: Lets do it in another PR later

**Review by YashP16 — APPROVED** [review]

File: contracts/e721-farms/camelotV3/CamelotV3Farm.sol [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+    /// @param _cooldownPeriod - cooldown period for locked deposits in days.
+    /// @dev _cooldownPeriod = 0 Disables lockup functionality for the farm.
+    /// @param _farmRegistry - Address of the Demeter Farm Registry.
+    /// @param _camelotPoolData - init data for CamelotV3 pool.
+    /// @param _rwdTokenData - init data for reward tokens.
+    /// @param _camelotV3Factory - Factory contract of Camelot V3.
+    /// @param _nftContract - NFT contract's address (NFPM).
+    /// @param _camelotUtils - address of our custom camelot utils contract.
+    /// @param _nfpmUtils - address of our custom camelot nonfungible position manager utils contract.
+    function initialize(
```
Comment (arcantheon) [inline contracts/e721-farms/camelotV3/CamelotV3Farm.sol]: Convert it to struct input like this: https://github.com/Sperax/Demeter-Protocol/blob/a8cfd61c98a9ebce56907a86f74913d0e623320f/contracts/e721-farms/uniswapV3/UniV3Farm.sol#L101

File: contracts/e721-farms/camelotV3/CamelotV3Farm.sol [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+
+        /// @dev Check if the token adheres to the tick range
+        if (positions.tickLower != tickLowerAllowed || positions.tickUpper != tickUpperAllowed) {
+            revert IncorrectTickRange();
+        }
+
+        return uint256(positions.liquidity);
+    }
+
+    function _validateTickRange(int24 _tickLower, int24 _tickUpper) private view {
```
Comment (arcantheon) [inline contracts/e721-farms/camelotV3/CamelotV3Farm.sol:280]: Please add natspec

**Review by arcantheon — CHANGES_REQUESTED** [review]
Added two comments

File: contracts/e721-farms/camelotV3/CamelotV3Farm.sol (hunk 7 above)
Comment (TechnoGeek01) [reply contracts/e721-farms/camelotV3/CamelotV3Farm.sol]: fixed in [edc3506](https://github.com/Sperax/Demeter-Protocol/pull/52/commits/edc350625863a6f48d4bacc0d8aa548d3dc18ddf)

File: contracts/e721-farms/camelotV3/CamelotV3Farm.sol (hunk 8 above)
Comment (TechnoGeek01) [reply contracts/e721-farms/camelotV3/CamelotV3Farm.sol:280]: fixed in [16b42db](https://github.com/Sperax/Demeter-Protocol/pull/52/commits/16b42dbb6f2883cd2b2ba03d56aef24dae39e2ab)

**Review by arcantheon — APPROVED** [review]

**Pull request merged**